In [1]:
from video_commands import *
main_video = "B0815_subtitle.mov"
width,height = get_media_dimensions(main_video)
print(width,height)

854 480


## check dimensions of a file

In [ ]:
width,height = get_media_dimensions("black.mov")
print(width,height)



## trim Video
outvideo = "test.mov"<br>
invideo = "B0815.mov"<br>
starttime,endtime = "0:10" , "0:18"<br>

In [ ]:
outvideo = "test_5min.mov"
invideo = "B0815_2flower.mov"
starttime,endtime = "00:00" , "5:00"
trim_video(invideo, starttime , endtime, outvideo)

## add a black still and insert in video<br>
input_video = "test_1min.mov"<br>
outfile = "test.mov"<br>
insert_black_stills(input_video=input_video,black_list=black_list,output_file=outfile,crf=18,preset="fast")



# # Text with colored parts
<font color = "red">add def, insert multiple new black stills using concat. Ask me to delete original mov</font>



In [ ]:
input_video = "B0815_2flower.mov"
text = "28年后的巨变"
#    create_black_still(input_video, text,5,10,output_file="black_still.mp4",font_name="Arial",font_size=150,font_color="&H00FFFFFF")
#    main_video = input("Enter path to main video: ")
#    text = input("Enter text for black still: ")
start_time = "17:00.5" #5
end_time = "17:03.5"   #10
black_file = "black.mov"
output_file = "B0815_3flower.mov"
crf = 10
preset = "fast"
create_black_still(input_video, text, start_time, end_time, black_file, font_name="Arial",font_size=130,font_color="&H00FFFFFF")
overlay_black_still_on_video(input_video, black_file,start_time,end_time,output_file=output_file,crf=crf,preset=preset)
    


In [ ]:
# insert new black stills 

In [ ]:
# insert new black stills 

## Audio - converting mp3 to wav<br>
mp3_file = "music/darkness.mp3"

In [ ]:
mp3_file = "music/darkness.mp3"
mp3_to_wav(mp3_file)

## Audio - adjusting volume<br>
wav_file = "music/darkness.wav" <br>
volume_factor = 0.1

In [ ]:
wav_file = "music/darkness.wav"
volume_factor = 0.1
adjust_wav_volume(wav_file, volume_factor)

## split video into sections<br>
sections = ["0:00-5:00", "5:00-8:00", "9:00-12:00"]<br>
split_video("xxx.mov", sections)<br>
output:test_5min_0005-0100.mov,test_5min_0100-0200.mov,test_5min_0220-0300.mov


In [ ]:
sections = ["0:05-1:00", "1:00-2:00", "2:20-3:00"]
input_file = "test_5min.mov"
split_video(input_file, sections)

## extract voice from a video and convert to srt file
lang can be None


In [ ]:
input_video, output_srt,lang = "test_1min.mov", "test_1min.srt","zh"
voice_to_srt(input_video, output_srt, lang=lang)#en,ja,k

## srt to ass

In [ ]:
srt_file = "test_1min.srt" #input file
ass_file = "test_1min.ass" #new file
srt_to_ass(srt_file, ass_file,width,height,fontname="宋体", fontsize=30)
burn_subtitles("test_1min.mov", "test_1min.ass", "output.mov", "Arial")


## Translation and Combine Different Translations for proofreading
<h3>only for srt!!!</h3>

In [ ]:
inputfile,outputfile,lang = "test_1min.srt","test_1min_jp.srt","JA" #lang = target language
deepl_translate_srt(inputfile, outputfile ,lang=lang)#EN-US KO zh JA

In [ ]:
inputfile = "test_1min.srt"
outputfile = "test_1min_jp.srt"
combined_file = "combined_chi_jp.srt"  #new file, oneline chinese one line japanese
combine_srt(inputfile,outputfile, combined_file)

## Add grid to a screenshot for mosaic
add_numbered_grid(<br>
    input_file,output_file,width,height,interval=100,line_color="red",<br>
    number_color="yellow",number_size=30,number_offset=5)

In [ ]:
screenshot,outscreenshot = "test2.png" ,"testout2.png"

In [ ]:
add_numbered_grid(
    screenshot,
    outscreenshot,
    width,
    height,
    interval=50,
    line_color="red",
    number_color="yellow",
    number_size=30,
    number_offset=5
)

# Apply Mosaic

## example: <br>
mosaic_list=[<br>
        [50, 50, 100, 50, 10, 0, 10],      <br>
        [800, 0, 200, 200, 15, "1:20.5", "3:00"]]  <br>
    # position of top-left corner<br>
    x=50,y=50,width=100,height=50<br>
    pixelization level=10<br>
    start seconds=0,end seconds=10<br>

In [ ]:
input_video,output_video = "test_5min.mov","test_5min_mosaic.mov"
mosaic_list = [[65, 50, 200, 170, 15, 0, 30],[width-260, 60, 285, 140, 15, "1:20.5", "2:30"]]
apply_mosaics(input_video, output_video, mosaic_list)

## srt to ass and burn in subtitles <br>
srt_file = "../中文字幕.srt" #input file <br>
ass_file = "中文字幕.ass" #new file <br>
input_video = "B0815_4letter.mov" <br>
output_video = "B0815_subtitle.mov" <br>
#srt_to_ass(srt_file, ass_file,width,height,fontname="宋体", fontsize=30) <br>
burn_subtitles(input_video, ass_file, output_video, "Arial",crf ="18",preset="veryfast") <br>

In [7]:
srt_file = "中文字幕.srt" #input file
ass_file = "中文字幕.ass" #new file
input_video = "B0815_4letter.mov"
output_video = "B0815_subtitle.mov"
srt_to_ass(srt_file, ass_file,width,height,fontname="宋体", fontsize=30)
burn_subtitles(input_video, ass_file, output_video, "Arial",crf ="10",preset="fast")

✅ Converted 中文字幕.srt → 中文字幕.ass successfully!


⚠️ B0815_subtitle.mov already exists. Delete it? (yes/no):  yes


Deleted old file: B0815_subtitle.mov
Found font directory: /System/Library/Fonts/Supplemental
Running: ffmpeg -i B0815_4letter.mov -vf subtitles=中文字幕.ass:force_style='FontName=Arial,PlayResX=854,PlayResY=480':fontsdir=/System/Library/Fonts/Supplemental -c:v libx264 -crf 10 -preset fast -c:a copy B0815_subtitle.mov


ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Subtitles burned into B0815_subtitle.mov


[out#0/mov @ 0x7f8b7df045c0] video:476980KiB audio:279413KiB subtitle:0KiB other streams:0KiB global headers:0KiB muxing overhead: 0.139225%
frame=40553 fps= 37 q=-1.0 Lsize=  757446KiB time=00:27:01.98 bitrate=3825.6kbits/s dup=0 drop=24 speed=1.48x elapsed=0:18:15.66    
[libx264 @ 0x7f8b7df04f00] frame I:249   Avg QP: 4.65  size: 74271
[libx264 @ 0x7f8b7df04f00] frame P:14467 Avg QP: 6.81  size: 22997
[libx264 @ 0x7f8b7df04f00] frame B:25837 Avg QP:10.10  size:  5312
[libx264 @ 0x7f8b7df04f00] consecutive B-frames:  1.8% 38.8%  3.1% 56.3%
[libx264 @ 0x7f8b7df04f00] mb I  I16..4: 24.9% 20.0% 55.1%
[libx264 @ 0x7f8b7df04f00] mb P  I16..4:  9.7%  9.4%  5.5%  P16..4: 30.4% 15.0% 10.8%  0.0%  0.0%    skip:19.2%
[libx264 @ 0x7f8b7df04f00] mb B  I16..4:  1.3%  1.6%  1.2%  B16..8: 20.2% 10.2%  2.1%  direct:12.7%  skip:50.8%  L0:43.9% L1:42.3% BI:13.8%
[libx264 @ 0x7f8b7df04f00] 8x8 transform intra:37.7% inter:25.6%
[libx264 @ 0x7f8b7df04f00] coded y,uvDC,uvAC intra: 54.3% 67.0% 50.7% inter:

In [ ]:
srt_file = "test_1min.srt" #input file
ass_file = "test_1min.ass" #new file
input_video = "test_1min.mov"
output_video = "test.mov"
srt_to_ass(srt_file, ass_file,width,height,fontname="宋体", fontsize=30)
burn_subtitles(input_video, ass_file, output_video, "Arial")

## End film
input "end_subtitle_txt" in the end film this way:<br>
编导演播：翰 光<br>
音乐：Chad Cannon<br>
照片提供： 小林熏<br>
         李晓斌<br>

The content of <b>end_subtitle_txt</b> is inserted into <b>end_subtitle_template</b> <br>
as <b>end_subtitle_ass</b> by <b><i>end_credits_ass</b></i>.<br>
Then <b><i>create_ending_film</b></i> is called to create <b>end_output</b> with <b>end_song</b> as background music.<br>
The dimension of  <b>end_output</b> is the same as the <b>main_video</b>. The background music has an adjustable volume<br>
and fade-in,fade-out time. <b>end_second</b> is the silent time before the end of the film and after the end of the song.
You can add a background video with an adjustable fade-out time. <br>
Remember to use <b><i>preprocess</b></i> to rescale it first.

In [4]:
end_subtitle_ass= "2508credits.ass"
end_subtitle_txt= "end_subtitle.txt"
end_subtitle_template = "end_subtitle_template.ass"
end_song = "music/end_song.wav"
end_output3 = "end_vol3.mov"
end_credits_ass(end_subtitle_template, end_subtitle_txt, end_subtitle_ass,first_start="0:00:01.40", 
                 each_duration_s=6.0,line_offset_s=0.60, fallback_duration_s=4.0)
create_ending_film(end_subtitle_ass, end_song, end_output3, width, height,
                       volume=3, end_second=5, fadein=3, fadeout=10,
                       bg_video=None,video_fadeout=4)

end_vol3.mov already exists. Delete it? [y/N]:  y


Deleted end_vol3.mov. Continuing...
FFmpeg command:
 ffmpeg -y -f lavfi -i color=c=black:s=854x480:d=58.0 -i music/end_song.wav -filter_complex color=c=black:s=854x480:d=58.0[v0];[v0]subtitles=2508credits.ass[v];[1:a]volume=3,afade=t=in:ss=0:d=3,afade=t=out:st=48.0:d=10[a] -map [v] -map [a] -c:v libx264 -c:a aac -shortest end_vol3.mov


ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

## Reencode and combine 
output_file="B0815_final1.5.mp4" not ok,must be <br>
output_file="B0815_final1.mp4", no "." in between

In [2]:

list_to_reencode = [
    ("beginning_ban_rescaled.mov", "beginning_ban_rescaled_reencoded.mov"),
    ("end_vol3.mov", "end_vol3_reencoded.mov")]
video_list = [list_to_reencode[0][1],main_video,list_to_reencode[1][1]]
#reencode_to_match(main_video, list_to_reencode)
combine_videos(video_list, output_file="B0815_final3.mov", crf=10,preset="fast")


Running: ffmpeg -f concat -safe 0 -i input.txt -c:v libx264 -crf 10 -preset fast -c:a aac -b:a 192k B0815_final3.mov


ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Combined video saved as B0815_final3.mov


In [2]:
#complicated method
time1,time2 = "13:09.8","13:14.9"
duration = parse_time(time2)-parse_time(time1)
print(duration)
overlay_video_split_and_combine(
    main_video="B0815_3flower.mov",
    overlay_video="video_folder/letter_muted.mp4",
    overlay_start="0:02",         # trim overlay from 1:01
    overlay_duration = duration,        # modify to the end
    main_insert_at= time1,    
    output_file="B0815_4letter.mov",# insert overlay at 5:00 of main
    crf=10,
    preset="fast",
    keep_intermediates=True
)



5.100000000000023

=========== ALL FFMPEG COMMANDS ===========
ffmpeg -y -i video_folder/letter_muted.mp4 -ss 2.000 -t 5.100 -an -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p tmp_overlay_trimmed.mp4
ffmpeg -y -i B0815_3flower.mov -ss 789.800 -t 5.100 -vn -acodec pcm_s16le -ar 44100 -ac 2 tmp_overlay_audio.wav
ffmpeg -y -i tmp_overlay_trimmed.mp4 -i tmp_overlay_audio.wav -c:v copy -c:a pcm_s16le tmp_overlay_with_audio.mp4
ffmpeg -y -i B0815_3flower.mov -t 789.800 -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le tmp_main_part1.mp4
ffmpeg -y -i B0815_3flower.mov -ss 794.900 -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le tmp_main_part2.mp4
ffmpeg -y -i tmp_main_part1.mp4 -i tmp_overlay_with_audio.mp4 -i tmp_main_part2.mp4 -filter_complex [0:v][0:a][1:v][1:a][2:v][2:a]concat=n=3:v=1:a=1[v][a] -map [v] -map [a] -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le -movflags +faststart B0815_4letter.mov


>>> Running FFmpeg:
ffmpeg -y -i vi

ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: tmp_overlay_trimmed.mp4

>>> Running FFmpeg:
ffmpeg -y -i B0815_3flower.mov -ss 789.800 -t 5.100 -vn -acodec pcm_s16le -ar 44100 -ac 2 tmp_overlay_audio.wav 



ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: tmp_overlay_audio.wav

>>> Running FFmpeg:
ffmpeg -y -i tmp_overlay_trimmed.mp4 -i tmp_overlay_audio.wav -c:v copy -c:a pcm_s16le tmp_overlay_with_audio.mp4 



ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: tmp_overlay_with_audio.mp4

>>> Running FFmpeg:
ffmpeg -y -i B0815_3flower.mov -t 789.800 -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le tmp_main_part1.mp4 



ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: tmp_main_part1.mp4

>>> Running FFmpeg:
ffmpeg -y -i B0815_3flower.mov -ss 794.900 -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le tmp_main_part2.mp4 



i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 34% 25% 11%  4%  6%  6%  5%  5%  4%
[libx264 @ 0x7fbbf52051c0] i8c dc,h,v,p: 61% 17% 18%  4%
[libx264 @ 0x7fbbf52051c0] Weighted P-Frames: Y:5.8% UV:0.9%
[libx264 @ 0x7fbbf52051c0] ref P L0: 80.8% 19.2%
[libx264 @ 0x7fbbf52051c0] ref B L0: 87.6% 12.4%
[libx264 @ 0x7fbbf52051c0] ref B L1: 98.8%  1.2%
[libx264 @ 0x7fbbf52051c0] kb/s:2514.94
ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtess

✅ Created: tmp_main_part2.mp4

>>> Running FFmpeg:
ffmpeg -y -i tmp_main_part1.mp4 -i tmp_overlay_with_audio.mp4 -i tmp_main_part2.mp4 -filter_complex [0:v][0:a][1:v][1:a][2:v][2:a]concat=n=3:v=1:a=1[v][a] -map [v] -map [a] -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le -movflags +faststart B0815_4letter.mov 



ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: B0815_4letter.mov
🎉 Done: B0815_4letter.mov


[out#0/mov @ 0x7fa24f006180] video:467261KiB audio:279413KiB subtitle:0KiB other streams:0KiB global headers:0KiB muxing overhead: 0.141361%
frame=40577 fps= 54 q=-1.0 Lsize=  747729KiB time=00:27:01.98 bitrate=3776.5kbits/s speed=2.17x elapsed=0:12:27.90    
[libx264 @ 0x7fa24f010c40] frame I:290   Avg QP: 4.50  size: 63063
[libx264 @ 0x7fa24f010c40] frame P:14541 Avg QP: 6.70  size: 22486
[libx264 @ 0x7fa24f010c40] frame B:25746 Avg QP: 9.97  size:  5174
[libx264 @ 0x7fa24f010c40] consecutive B-frames:  1.9% 39.7%  2.6% 55.8%
[libx264 @ 0x7fa24f010c40] mb I  I16..4: 26.0% 19.5% 54.5%
[libx264 @ 0x7fa24f010c40] mb P  I16..4:  9.7%  9.7%  5.3%  P16..4: 30.4% 14.9% 10.8%  0.0%  0.0%    skip:19.2%
[libx264 @ 0x7fa24f010c40] mb B  I16..4:  1.3%  1.6%  1.2%  B16..8: 20.5% 10.0%  2.0%  direct:12.7%  skip:50.9%  L0:43.7% L1:42.4% BI:13.9%
[libx264 @ 0x7fa24f010c40] 8x8 transform intra:38.2% inter:26.1%
[libx264 @ 0x7fa24f010c40] coded y,uvDC,uvAC intra: 54.4% 66.5% 50.0% inter: 16.9% 29.4% 1

In [2]:
#complicated method
time1,time2 = "13:09.8","13:14.9"
duration = parse_time(time2)-parse_time(time1)
print(duration)
overlay_video_split_and_combine(
    main_video="B0815_3flower.mov",
    overlay_video="video_folder/letter_muted.mp4",
    overlay_start="0:02",         # trim overlay from 1:01
    overlay_duration = duration,        # modify to the end
    main_insert_at= time1,    
    output_file="B0815_4letter.mov",# insert overlay at 5:00 of main
    crf=10,
    preset="fast",
    keep_intermediates=True
)



5.100000000000023

=========== ALL FFMPEG COMMANDS ===========
ffmpeg -y -i video_folder/letter_muted.mp4 -ss 2.000 -t 5.100 -an -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p tmp_overlay_trimmed.mp4
ffmpeg -y -i B0815_3flower.mov -ss 789.800 -t 5.100 -vn -acodec pcm_s16le -ar 44100 -ac 2 tmp_overlay_audio.wav
ffmpeg -y -i tmp_overlay_trimmed.mp4 -i tmp_overlay_audio.wav -c:v copy -c:a pcm_s16le tmp_overlay_with_audio.mp4
ffmpeg -y -i B0815_3flower.mov -t 789.800 -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le tmp_main_part1.mp4
ffmpeg -y -i B0815_3flower.mov -ss 794.900 -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le tmp_main_part2.mp4
ffmpeg -y -i tmp_main_part1.mp4 -i tmp_overlay_with_audio.mp4 -i tmp_main_part2.mp4 -filter_complex [0:v][0:a][1:v][1:a][2:v][2:a]concat=n=3:v=1:a=1[v][a] -map [v] -map [a] -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le -movflags +faststart B0815_4letter.mov


>>> Running FFmpeg:
ffmpeg -y -i vi

ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: tmp_overlay_trimmed.mp4

>>> Running FFmpeg:
ffmpeg -y -i B0815_3flower.mov -ss 789.800 -t 5.100 -vn -acodec pcm_s16le -ar 44100 -ac 2 tmp_overlay_audio.wav 



ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: tmp_overlay_audio.wav

>>> Running FFmpeg:
ffmpeg -y -i tmp_overlay_trimmed.mp4 -i tmp_overlay_audio.wav -c:v copy -c:a pcm_s16le tmp_overlay_with_audio.mp4 



ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: tmp_overlay_with_audio.mp4

>>> Running FFmpeg:
ffmpeg -y -i B0815_3flower.mov -t 789.800 -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le tmp_main_part1.mp4 



ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: tmp_main_part1.mp4

>>> Running FFmpeg:
ffmpeg -y -i B0815_3flower.mov -ss 794.900 -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le tmp_main_part2.mp4 



i4 v,h,dc,ddl,ddr,vr,hd,vl,hu: 34% 25% 11%  4%  6%  6%  5%  5%  4%
[libx264 @ 0x7fbbf52051c0] i8c dc,h,v,p: 61% 17% 18%  4%
[libx264 @ 0x7fbbf52051c0] Weighted P-Frames: Y:5.8% UV:0.9%
[libx264 @ 0x7fbbf52051c0] ref P L0: 80.8% 19.2%
[libx264 @ 0x7fbbf52051c0] ref B L0: 87.6% 12.4%
[libx264 @ 0x7fbbf52051c0] ref B L1: 98.8%  1.2%
[libx264 @ 0x7fbbf52051c0] kb/s:2514.94
ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtess

✅ Created: tmp_main_part2.mp4

>>> Running FFmpeg:
ffmpeg -y -i tmp_main_part1.mp4 -i tmp_overlay_with_audio.mp4 -i tmp_main_part2.mp4 -filter_complex [0:v][0:a][1:v][1:a][2:v][2:a]concat=n=3:v=1:a=1[v][a] -map [v] -map [a] -c:v libx264 -crf 10 -preset fast -pix_fmt yuv420p -c:a pcm_s16le -movflags +faststart B0815_4letter.mov 



ffmpeg version 8.0 Copyright (c) 2000-2025 the FFmpeg developers
  built with Apple clang version 15.0.0 (clang-1500.1.0.2.5)
  configuration: --prefix=/usr/local/Cellar/ffmpeg/8.0 --enable-shared --enable-pthreads --enable-version3 --cc=clang --host-cflags= --host-ldflags='-Wl,-ld_classic' --enable-ffplay --enable-gnutls --enable-gpl --enable-libaom --enable-libaribb24 --enable-libbluray --enable-libdav1d --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librav1e --enable-librist --enable-librubberband --enable-libsnappy --enable-libsrt --enable-libssh --enable-libsvtav1 --enable-libtesseract --enable-libtheora --enable-libvidstab --enable-libvmaf --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libxvid --enable-lzma --enable-libfontconfig --enable-libfreetype --enable-frei0r --enable-libass --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libspeex --enabl

✅ Created: B0815_4letter.mov
🎉 Done: B0815_4letter.mov


[out#0/mov @ 0x7fa24f006180] video:467261KiB audio:279413KiB subtitle:0KiB other streams:0KiB global headers:0KiB muxing overhead: 0.141361%
frame=40577 fps= 54 q=-1.0 Lsize=  747729KiB time=00:27:01.98 bitrate=3776.5kbits/s speed=2.17x elapsed=0:12:27.90    
[libx264 @ 0x7fa24f010c40] frame I:290   Avg QP: 4.50  size: 63063
[libx264 @ 0x7fa24f010c40] frame P:14541 Avg QP: 6.70  size: 22486
[libx264 @ 0x7fa24f010c40] frame B:25746 Avg QP: 9.97  size:  5174
[libx264 @ 0x7fa24f010c40] consecutive B-frames:  1.9% 39.7%  2.6% 55.8%
[libx264 @ 0x7fa24f010c40] mb I  I16..4: 26.0% 19.5% 54.5%
[libx264 @ 0x7fa24f010c40] mb P  I16..4:  9.7%  9.7%  5.3%  P16..4: 30.4% 14.9% 10.8%  0.0%  0.0%    skip:19.2%
[libx264 @ 0x7fa24f010c40] mb B  I16..4:  1.3%  1.6%  1.2%  B16..8: 20.5% 10.0%  2.0%  direct:12.7%  skip:50.9%  L0:43.7% L1:42.4% BI:13.9%
[libx264 @ 0x7fa24f010c40] 8x8 transform intra:38.2% inter:26.1%
[libx264 @ 0x7fa24f010c40] coded y,uvDC,uvAC intra: 54.4% 66.5% 50.0% inter: 16.9% 29.4% 1

In [2]:
make_rescaled_image(width, height, "班和保人.png", "test.png")

FFmpeg command: ffmpeg -y -i 班和保人.png -vf scale=669:480 test.png
Successfully created test.png


'test.png'

In [ ]:
#easy overlay video1(from a certain time) to video2(from time1 to time2)
video1_path = "letter.MP4"  #overlay video
video2_path = "B0815_3flower.mov"   #main video
overlay_start_main = "13:09.8" #"13:09.8","13:14.9"
overlay_end_main = "13:14.9"
extract_start_video1 = 2
output_filename = "test.mov"
overlay_video(
        video1_path,
        video2_path,
        overlay_start_main,
        overlay_end_main,
        extract_start_video1,
        output_filename,
        crf=30,
        preset="ultrafast"
    )


Extracting clip from overlay video...
FFmpeg command: ffmpeg -ss 2.0 -i letter.MP4 -t 5.100000000000023 -c copy -y /var/folders/9g/509r85b95jz43885pc6n7zlr0000gn/T/tmp3zwj1orw/extracted_clip.mp4
Removing audio from extracted clip...
FFmpeg command: ffmpeg -i /var/folders/9g/509r85b95jz43885pc6n7zlr0000gn/T/tmp3zwj1orw/extracted_clip.mp4 -an -y /var/folders/9g/509r85b95jz43885pc6n7zlr0000gn/T/tmp3zwj1orw/muted_clip.mp4


1.automatic rescale of images and videos in the code or do reencoding in the preprocessing<br>
2.insert multiple black stills<br>
3.insert multiple images and videos<br>
4.overlay a single image, or list image<br>
overlay multiple videos<br>
5.overlay_black_still not working in github<br>
6. add preprocessing<br>
7. 更改片尾字幕resolution<br>
8. convert the following to python:<br>
ffmpeg -i B0815_1mosaic_better_quality.mov -i 黄花菜地_muted.mov -filter_complex "<br>
[0:v]trim=0:1601,setpts=PTS-STARTPTS[part1];<br>
color=c=black:size=854x480<br>
:d=21[blackpart];<br>
[1:v]trim=0:21,setpts=PTS-STARTPTS,format=rgba,fade=t=out:st=19:d=2:alpha=1[ov];<br>
[blackpart][ov]overlay=x=(main_w-overlay_w)/2:y=(main_h-overlay_h)/2:shortest=1[part2];<br>
[part1][part2]concat=n=2:v=1:a=0[vout]<br>
" -map "[vout]" -map 0:a -c:v libx264 -pix_fmt yuv420p -preset veryfast -crf 18 -c:a copy B0815_2flower.mov<br>
9. convert it to python for overlay. start overlay time is 789.8s, upto 794.9 <br>
ffmpeg -i B0815_subtitle.mov -i video_folder/letter_muted.mp4 -filter_complex \
"[1:v]setpts=PTS+789.8/TB[v1]; \
 [0:v][v1]overlay=0:0:enable='between(t,789.8,794.9)'" \
-c:a copy -c:v libx264 -crf 30 -preset ultrafast output.mov



<h3> ***continue here , burn in subtitle </h3>